### PR TITLE
added PositionTrainer for is2rs direct 

### PR DESCRIPTION
**Pull request by AdeeshKolluru**

### CODE REVIEW

File: ocpmodels/trainers/position_trainer.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+
+            # perform loss normalization before backprop
+            train_loss_position_normalized = train_loss_position_unnormalized * (
+                distutils.get_world_size() / train_loss_position_normalizer
+            )
+            loss = train_loss_position_normalized
+        
+        else:
+            try:
+                loss = self.criterion(out["position"], target_normed)
```
Comment (mshuaibii) [inline ocpmodels/trainers/position_trainer.py:377]: Let's add the logic for training on free vs. fixed atoms here: https://github.com/ulissigroup/ocp/blob/55590774fb4fe9220bdb30e53645c81dd4869cb9/ocpmodels/trainers/forces_trainer.py#L533-L543.

File: ocpmodels/trainers/position_trainer.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+            is_vis=is_vis,
+            print_every=print_every,
+            seed=seed,
+            logger=logger,
+            logger_project=logger_project,
+            logger_entity=logger_entity,
+            local_rank=local_rank,
+            amp=amp,
+            cpu=cpu,
+            name="is2re",
```
Comment (mshuaibii) [inline ocpmodels/trainers/position_trainer.py:94]: change this to is2rs"

Comment (mshuaibii) [reply ocpmodels/trainers/position_trainer.py:94]: the metrics currently associated with "is2rs" are found here: https://github.com/Open-Catalyst-Project/ocp/blob/899de1c1f46f44bf99e3ccd190b9a37795b60b67/ocpmodels/modules/evaluator.py#L43.

We probably want to remove `energy_mae` and replace it with `positions_mae`: https://github.com/Open-Catalyst-Project/ocp/blob/899de1c1f46f44bf99e3ccd190b9a37795b60b67/ocpmodels/modules/evaluator.py#L154-L155

File: ocpmodels/trainers/position_trainer.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+                predictions["position"] = out["position"].detach()
+                return predictions
+            
+        predictions["id"].extend([str(i) for i in batch[0].sid.tolist()])
+        predictions["position"].extend(out["position"].numpy())
+        self.save_results(predictions, results_file, keys=["position"])
+        return predictions
+
+    def train(self):
+        self.best_val_mae = 1e9
```
Comment (mshuaibii) [inline ocpmodels/trainers/position_trainer.py:250]: I would replace this logic with https://github.com/Open-Catalyst-Project/ocp/blob/899de1c1f46f44bf99e3ccd190b9a37795b60b67/ocpmodels/trainers/forces_trainer.py#L337-L340. The primary metric for `is2rs` is ADwT (higher is better).

File: ocpmodels/trainers/position_trainer.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+                        step=epoch * len(self.train_loader) + i + 1,
+                        split="train",
+                    )
+
+            self.scheduler.step()
+            torch.cuda.empty_cache()
+
+            if self.val_loader is not None:
+                val_metrics = self.validate(split="val", epoch=epoch)
+                if (
```
Comment (mshuaibii) [inline ocpmodels/trainers/position_trainer.py:300]: Same here - I'd replace this logic with: https://github.com/Open-Catalyst-Project/ocp/blob/899de1c1f46f44bf99e3ccd190b9a37795b60b67/ocpmodels/trainers/forces_trainer.py#L395-L408 so it saves val metrics accordingly depending on what primary metric is used.

**Review by mshuaibii — CHANGES_REQUESTED** [review]
Looks good - few changes before we merge this.

File: ocpmodels/models/dimenet_plus_plus.py [hunk 5]
```diff
@@ -333,7 +333,8 @@ def __init__(
         bond_feat_dim,  # not used
         num_targets,
         use_pbc=True,
-        regress_forces=True,
+        regress_forces=False,
```
Comment (mshuaibii) [inline ocpmodels/models/dimenet_plus_plus.py:336]: this isn't backwards compatible with the configs - if we plan on merging this to master keep this as true and make `regress_positions=False`. In it's corresponding config you can update these.

File: ocpmodels/models/dimenet_plus_plus.py [hunk 6]
```diff
@@ -372,7 +379,7 @@ def __init__(
 
     def forward(self, data):
         pos = data.pos
-        if self.regress_forces:
+        if self.regress_forces or self.regress_position:
```
Comment (mshuaibii) [inline ocpmodels/models/dimenet_plus_plus.py:382]: we don't need to keep track of gradients here - this is only needed if we plan on differentiating twice through the model (forces)

File: ocpmodels/models/dimenet_plus_plus.py [hunk 7]
```diff
@@ -439,21 +446,27 @@ def forward(self, data):
...
-                    pos,
-                    grad_outputs=torch.ones_like(energy),
-                    create_graph=True,
-                )[0]
-            )
-            return energy, forces
+        
+        if self.regress_position:
+            position = P #P.sum(dim=0) if batch is None else scatter(P, batch, dim=0)
+            #import pdb; pdb.set_trace()
```
Comment (mshuaibii) [inline ocpmodels/models/dimenet_plus_plus.py:452]: remove the debug comments here

File: ocpmodels/modules/evaluator.py [hunk 8]
```diff
@@ -41,23 +41,26 @@ class Evaluator:
...
         "is2rs": ["average_distance_within_threshold", "energy_mae"],
+        "is2rs_direct": ["average_distance_within_threshold", "positions_mae", "positions_mse"],
         "is2re": ["energy_mae", "energy_mse", "energy_within_threshold"],
     }
 
     task_attributes = {
         "s2ef": ["energy", "forces", "natoms"],
-        "is2rs": ["positions", "cell", "pbc", "energy", "natoms"],
+        "is2rs": ["position", "cell", "pbc", "energy", "natoms"],
+        "is2rs_direct": ["position"],
```
Comment (mshuaibii) [inline ocpmodels/modules/evaluator.py:51]: lets keep this all `positions`

**Review by mshuaibii — CHANGES_REQUESTED** [review]

File: ocpmodels/models/dimenet_plus_plus.py [hunk 9]
```diff
@@ -334,6 +334,7 @@ def __init__(
         num_targets,
         use_pbc=True,
         regress_forces=True,
+        regress_position=True,
```
Comment (mshuaibii) [inline ocpmodels/models/dimenet_plus_plus.py:337]: default this to False - will break other configs. This is taken care of here: https://github.com/ulissigroup/ocp/pull/1/files#diff-9339613c7385303c958c59afb6bdbce973badd44362355ef085e7b4aa438ead3R16 so you're fine.

**Review by mshuaibii — CHANGES_REQUESTED** [review]

**Review by mshuaibii — APPROVED** [review]
LGTM - thanks!
